packages/frontend/workspace/src/providers/sync/peer.ts

```diff
@@ -0,0 +1,379 @@
+    if (s !== this._status) {
+      this.logger.debug('status change', SyncPeerStatus[s]);
+      this._status = s;
+      this.onStatusChange.emit(s);
+    }
+  }
+
+  get status() {
+    return this._status;
+  }
+
+  /**
+   * stop sync
+   *
+   * SyncPeer is one-time use, this peer should be discarded after call stop().
+   */
+  stop() {
+    this.logger.debug('peer stop');
+    this.abort.abort('manually-stop');
+  }
+
+  /**
+   * auto retry after 5 seconds if sync failed
+   */
+  async syncRetryLoop(abort: AbortSignal) {
+    while (abort.aborted === false) {
+      try {
+        await this.sync(abort);
+      } catch (err) {
+        if (err === 'manually-stop') {
+          return;
+        }
+
+        this.logger.error('sync error', err);
+      }
+      try {
+        this.logger.error('retry after 5 seconds');
+        this.status = SyncPeerStatus.Retrying;
+        await Promise.race([
+          new Promise<void>(resolve => {
+            setTimeout(resolve, 5 * 1000);
+          }),
+          new Promise((_, reject) => {
+            // exit if manually stopped
+            if (abort.aborted) {
+              reject(abort.reason);
+            }
+            abort.addEventListener('abort', () => {
+              reject(abort.reason);
+            });
+          }),
+        ]);
+      } catch (err) {
+        if (err === 'manually-stop') {
+          return;
+        }
+
+        // should never reach here
+        throw err;
+      }
+    }
+  }
+
+  private state: {
+    connectedDocs: Map<string, Doc>;
+    pushUpdatesQueue: AsyncQueue<{
+      docId: string;
+      data: Uint8Array;
+    }>;
+    pullUpdatesQueue: AsyncQueue<{
+      docId: string;
+      data: Uint8Array;
+    }>;
+    subdocsLoadQueue: AsyncQueue<Doc>;
+  } = {
+    connectedDocs: new Map(),
+    pushUpdatesQueue: new AsyncQueue(),
+    pullUpdatesQueue: new AsyncQueue(),
+    subdocsLoadQueue: new AsyncQueue(),
+  };
+
+  /**
+   * main synchronization logic
+   */
+  async sync(abortOuter: AbortSignal) {
+    const abortInner = new AbortController();
+
+    abortOuter.addEventListener('abort', reason => {
+      abortInner.abort(reason);
+    });
+
+    let dispose: (() => void) | null = null;
+    try {
+      // start listen storage updates
+      dispose = await this.storage.subscribe(
+        this.handleStorageUpdates,
+        reason => {
+          // abort if storage disconnect, should trigger retry loop
+          abortInner.abort('subscribe disconnect:' + reason);
+        }
+      );
+      throwIfAborted(abortInner.signal);
+
+      // Step 1: load root doc
+      this.status = SyncPeerStatus.LoadingRootDoc;
+
+      await this.connectDoc(this.rootDoc, abortInner.signal);
+
+      this.status = SyncPeerStatus.LoadingSubDoc;
+
+      // Step 2: load subdocs
+      this.state.subdocsLoadQueue.push(
+        ...Array.from(this.rootDoc.getSubdocs())
+      );
+
+      this.rootDoc.on('subdocs', this.handleSubdocsUpdate);
```

Comment:
I think it should be put before sync root doc? Will the subdocs event be fired after that?

Reply:
After sync root doc, user can create new pages, and also possible create new subdocs by remotes.

We get all subdocs here and listen for new subdocs after, We can guarantee that all subdocs are in the load queue.